pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,284 @@
+	var srcIP, dstIP string
+	var srcPort, dstPort uint16
+	var proto uint8
+	// record has complete 5-tuple information (IPv4)
+	infoElementMap := record.GetInfoElementMap()
+	if containsInfoElement(infoElementMap, "sourceIPv4Address") && containsInfoElement(infoElementMap, "destinationIPv4Address") && containsInfoElement(infoElementMap, "sourceTransportPort") && containsInfoElement(infoElementMap, "destinationTransportPort") && containsInfoElement(infoElementMap, "protocolIdentifier") {
+		srcIPAddr, ok := infoElementMap["sourceIPv4Address"].Value.(net.IP)
+		if !ok {
+			return Tuple{}, fmt.Errorf("sourceIPv4Address is not in correct format.")
+		}
+		srcIP = srcIPAddr.String()
+		dstIPAddr, ok := infoElementMap["destinationIPv4Address"].Value.(net.IP)
+		if !ok {
+			return Tuple{}, fmt.Errorf("destinationIPv4Address is not in correct format.")
+		}
+		dstIP = dstIPAddr.String()
+		srcPortNum, ok := infoElementMap["sourceTransportPort"].Value.(uint16)
+		if !ok {
+			return Tuple{}, fmt.Errorf("sourceTransportPort is not in correct format.")
+		}
+		srcPort = srcPortNum
+		dstPortNum, ok := infoElementMap["destinationTransportPort"].Value.(uint16)
+		if !ok {
+			return Tuple{}, fmt.Errorf("destinationTransportPort is not in correct format.")
+		}
+		dstPort = dstPortNum
+		protoNum, ok := infoElementMap["protocolIdentifier"].Value.(uint8)
+		if !ok {
+			return Tuple{}, fmt.Errorf("protocolIdentifier is not in correct format.")
+		}
+		proto = protoNum
+	} else if containsInfoElement(infoElementMap, "sourceIPv6Address") && containsInfoElement(infoElementMap, "destinationIPv6Address") && containsInfoElement(infoElementMap, "sourceTransportPort") && containsInfoElement(infoElementMap, "destinationTransportPort") && containsInfoElement(infoElementMap, "protocolIdentifier") {
+		srcIPAddr, ok := infoElementMap["sourceIPv6Address"].Value.(net.IP)
+		if !ok {
+			return Tuple{}, fmt.Errorf("sourceIPv6Address is not in correct format.")
+		}
+		srcIP = net.IP(srcIPAddr).String()
+		dstIPAddr, ok := infoElementMap["destinationIPv6Address"].Value.(net.IP)
+		if !ok {
+			return Tuple{}, fmt.Errorf("destinationIPv6Address is not in correct format.")
+		}
+		dstIP = net.IP(dstIPAddr).String()
+		srcPortNum, ok := infoElementMap["sourceTransportPort"].Value.(uint16)
+		if !ok {
+			return Tuple{}, fmt.Errorf("sourceTransportPort is not in correct format.")
+		}
+		srcPort = srcPortNum
+		dstPortNum, ok := infoElementMap["destinationTransportPort"].Value.(uint16)
+		if !ok {
+			return Tuple{}, fmt.Errorf("destinationTransportPort is not in correct format.")
+		}
+		dstPort = dstPortNum
+		protoNum, ok := infoElementMap["protocolIdentifier"].Value.(uint8)
+		if !ok {
+			return Tuple{}, fmt.Errorf("protocolIdentifier is not in correct format.")
+		}
+		proto = protoNum
+	} else {
+		return Tuple{}, fmt.Errorf("missing 5-tuple value(s) in the record.")
+	}
+	return Tuple{srcIP, dstIP, proto, srcPort, dstPort}, nil
+}
```

Comment:
We may consider simplifying getTupleFromRecord() as the following, although the error report for the address part may not be precious.

```
func (b *baseRecord) GetInfoElementValue(name string, dataType entities.IEDataType) (interface{}, error) {
        if element, exist := b.elementsMap[name]; exist {
                return entities.DecodeToIEDataType(dataType, element.Value)
                /*                                                                                             
                 * If element.Element.DataType is populated with correct type,
                 * then we don't need to pass dataType as an input. We can just do                                                                              
                 *    return entities.DecodeToIEDataType(element.Element.DataType, element.Value)              
                 */
        } else {
                return nil, fmt.Errorf("missing %s in the record.", name)
        }
}

func getTupleFromRecord(record entities.Record) (Tuple, error) {
        srcPort, err := record.GetInfoElementValue("sourceTransportPort", entities.Unsigned16)
        if err {
                return Tuple{}, err
        }

        dstPort, err := record.GetInfoElementValue("destinationTransportPort", entities.Unsigned16)
        if err {
                return Tuple{}, err
        }

        proto, err := record.GetInfoElementValue("destinationTransportPort", entities.Unsigned8)
        if err {
                return Tuple{}, err
        }

        srcIPAddr, err := record.GetInfoElementValue("sourceIPv4Address", entities.Ipv4Address)
        if err {
                srcIPAddr, err = record.GetInfoElementValue("sourceIPv6Address", entities.Ipv6Address)
        }
        if err {
                return Tuple{}, err
        }
        srcIP := srcIPAddr.String()

        dstIPAddr, err := record.GetInfoElementValue("destinationIPv4Address", entities.Ipv4Address)
        if err {
                dstIPAddr, err = record.GetInfoElementValue("destinationIPv6Address", entities.Ipv6Address)
        }
        if err {
                return Tuple{}, err
        }
        dstIP := dstIPAddr.String()

        return Tuple{srcIP, dstIP, proto, srcPort, dstPort}, nil
}
```

Reply:
@zyiou My main point was not having two interfaces and just to have one. You had both `GetInfoElementValue` and `ContainsInfoElement`.

Above modification to`GetInfoElementValue` looks good. However, I suggest to return `exists` bool rather than error like (interface{}, bool, error). Caller can consume `exists`.

 A note in above code: It's better to return `nil, err` rather than `Tuple{}, err`. We can save on malloc.

Reply:
Got it. will update that. Thanks!